# MLP-0: Data discovery - Telco Customer Churn

**Goal:** assess data quality, identify which customer segments churn most (input for the Retention team), and define the feature set for the training pipeline (MLP-1).

**Data:** [Telco Customer Churn](https://www.kaggle.com/datasets/blastchar/telco-customer-churn) (IBM sample data), 7,043 customers × 21 columns. Place the CSV in `data/raw/` before running.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [3]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [4]:
df.shape

(7043, 21)

In [5]:
df.dtypes

customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object

In [6]:
df["customerID"].duplicated().sum()

np.int64(0)

In [7]:
df.drop(columns="customerID").duplicated(keep=False).sum()

np.int64(42)

In [8]:
duplicates_mask = df.drop(columns="customerID").duplicated(keep=False)
df[duplicates_mask].sort_values("MonthlyCharges")

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
542,2866-IKBTM,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.55,19.55,No
1491,8605-ITULD,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.55,19.55,No
6774,0970-QXPXW,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.65,19.65,No
5170,7721-DVEKZ,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.65,19.65,No
6706,7878-RTCZG,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.90,19.9,No
4817,6654-QGBZZ,Female,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,19.90,19.9,No
662,3030-ZKIWL,Male,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,20.05,20.05,No
5736,8749-CLJXC,Male,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,20.05,20.05,No
3754,7096-UCLNH,Male,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,No,Mailed check,20.05,20.05,No
3679,1240-KNSEZ,Male,0,No,No,1,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Mailed check,20.10,20.1,Yes


In [9]:
test_columns = df.drop(columns="customerID").select_dtypes(include="str").columns
test_columns

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn'],
      dtype='str')

In [10]:
for col in df[test_columns]:
    print(col)
    print(df[col].nunique())
    print(df[col].value_counts())
    print("---------")

gender
2
gender
Male      3555
Female    3488
Name: count, dtype: int64
---------
Partner
2
Partner
No     3641
Yes    3402
Name: count, dtype: int64
---------
Dependents
2
Dependents
No     4933
Yes    2110
Name: count, dtype: int64
---------
PhoneService
2
PhoneService
Yes    6361
No      682
Name: count, dtype: int64
---------
MultipleLines
3
MultipleLines
No                  3390
Yes                 2971
No phone service     682
Name: count, dtype: int64
---------
InternetService
3
InternetService
Fiber optic    3096
DSL            2421
No             1526
Name: count, dtype: int64
---------
OnlineSecurity
3
OnlineSecurity
No                     3498
Yes                    2019
No internet service    1526
Name: count, dtype: int64
---------
OnlineBackup
3
OnlineBackup
No                     3088
Yes                    2429
No internet service    1526
Name: count, dtype: int64
---------
DeviceProtection
3
DeviceProtection
No                     3095
Yes                    2422
No in

In [11]:
service_columns = ["OnlineSecurity", "OnlineBackup", "DeviceProtection", "TechSupport",
                   "StreamingTV", "StreamingMovies"]

for col in service_columns:
    print(col)
    display(pd.crosstab(df["InternetService"], df[col]))
    print("-----")

pd.crosstab(df["PhoneService"], df["MultipleLines"])

OnlineSecurity


OnlineSecurity,No,No internet service,Yes
InternetService,,,
DSL,1241,0,1180
Fiber optic,2257,0,839
No,0,1526,0


-----
OnlineBackup


OnlineBackup,No,No internet service,Yes
InternetService,,,
DSL,1335,0,1086
Fiber optic,1753,0,1343
No,0,1526,0


-----
DeviceProtection


DeviceProtection,No,No internet service,Yes
InternetService,,,
DSL,1356,0,1065
Fiber optic,1739,0,1357
No,0,1526,0


-----
TechSupport


TechSupport,No,No internet service,Yes
InternetService,,,
DSL,1243,0,1178
Fiber optic,2230,0,866
No,0,1526,0


-----
StreamingTV


StreamingTV,No,No internet service,Yes
InternetService,,,
DSL,1464,0,957
Fiber optic,1346,0,1750
No,0,1526,0


-----
StreamingMovies


StreamingMovies,No,No internet service,Yes
InternetService,,,
DSL,1440,0,981
Fiber optic,1345,0,1751
No,0,1526,0


-----


MultipleLines,No,No phone service,Yes
PhoneService,,,
No,0,682,0
Yes,3390,0,2971


In [12]:
df.loc[df["TotalCharges"].str.strip().eq(""), "tenure"].unique()

array([0])

In [13]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [14]:
df["TotalCharges"].isna().sum()

np.int64(11)

In [15]:
df["TotalCharges"] = df["TotalCharges"].fillna(0)

In [16]:
df["TotalCharges"].isna().sum()


np.int64(0)

In [17]:
df["Churn"] = df["Churn"].map({"Yes": 1, "No": 0})

In [18]:
df["Churn"].value_counts(dropna=False)

Churn
0    5174
1    1869
Name: count, dtype: int64

In [19]:
df[["tenure", "MonthlyCharges", "TotalCharges"]].describe()

,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000
mean,32.371149,64.761692,2279.734304
std,24.559481,30.090047,2266.794470
min,0.000000,18.250000,0.000000
25%,9.000000,35.500000,398.550000
50%,29.000000,70.350000,1394.550000
75%,55.000000,89.850000,3786.600000
max,72.000000,118.750000,8684.800000


In [20]:
categorical_columns = ["gender", "SeniorCitizen", "Partner", "Dependents", "PhoneService", "MultipleLines", "InternetService", "OnlineSecurity", "OnlineBackup", "DeviceProtection", "TechSupport", "StreamingTV", "StreamingMovies", "Contract", "PaperlessBilling", "PaymentMethod"]
len(categorical_columns)

16

In [21]:
for col in categorical_columns:
    print(col)
    print(df.groupby(col)["Churn"].mean().sort_values(ascending=False))
    print("-----")

gender
gender
Female    0.269209
Male      0.261603
Name: Churn, dtype: float64
-----
SeniorCitizen
SeniorCitizen
1    0.416813
0    0.236062
Name: Churn, dtype: float64
-----
Partner
Partner
No     0.329580
Yes    0.196649
Name: Churn, dtype: float64
-----
Dependents
Dependents
No     0.312791
Yes    0.154502
Name: Churn, dtype: float64
-----
PhoneService
PhoneService
Yes    0.267096
No     0.249267
Name: Churn, dtype: float64
-----
MultipleLines
MultipleLines
Yes                 0.286099
No                  0.250442
No phone service    0.249267
Name: Churn, dtype: float64
-----
InternetService
InternetService
Fiber optic    0.418928
DSL            0.189591
No             0.074050
Name: Churn, dtype: float64
-----
OnlineSecurity
OnlineSecurity
No                     0.417667
Yes                    0.146112
No internet service    0.074050
Name: Churn, dtype: float64
-----
OnlineBackup
OnlineBackup
No                     0.399288
Yes                    0.215315
No internet service    0.

In [22]:
numeric_columns = ["tenure", "MonthlyCharges", "TotalCharges"]
df.groupby("Churn")[numeric_columns].agg(["median", "mean"])

tenure            MonthlyCharges            TotalCharges             
      median       mean         median       mean       median         mean
Churn                                                                      
0       38.0  37.569965         64.425  61.265124     1679.525  2549.911442
1       10.0  17.979133         79.650  74.441332      703.550  1531.796094

In [23]:
df["tenure_group"] = pd.cut(
    df["tenure"],
    bins=[0, 12, 24, 48, 72],
    labels=["0-12", "13-24", "25-48", "49-72"],
    include_lowest=True,
)
df.groupby("tenure_group")["Churn"].agg(["mean", "count"])

,mean,count
tenure_group,,
0-12,0.474382,2186
13-24,0.287109,1024
25-48,0.203890,1594
49-72,0.095132,2239


In [24]:
df[numeric_columns].corr()

,tenure,MonthlyCharges,TotalCharges
tenure,1.000000,0.247900,0.826178
MonthlyCharges,0.247900,1.000000,0.651174
TotalCharges,0.826178,0.651174,1.000000


## Summary

### 1. Data quality
- **Shape:** 7,043 customers × 21 columns. No missing values except `TotalCharges` (see below).
- **Duplicates:** no duplicate `customerID`s. Excluding the ID, 42 rows share an identical profile with at least one other row. All of them have `tenure = 1` (one bill, few possible feature combinations), so they are treated as **distinct customers with the same profile** and kept.
- **`TotalCharges`:** numeric values stored as text. 11 values are blank strings (whitespace, not `NaN`, so `isna()` does not detect them). All 11 belong to customers with `tenure = 0`, i.e. new customers who have not been billed yet, so they are **filled with 0** (a true value, not an imputation). Converted with `pd.to_numeric(errors="coerce")`, and the number of `NaN`s was verified to be exactly 11 before filling.
- **Consistency checks:** `"No internet service"` appears in all 6 internet add-on columns **if and only if** `InternetService = "No"` (1,526 customers). `"No phone service"` in `MultipleLines` matches `PhoneService = "No"` (682). The data is consistent, but these values are **redundant**: the same information is repeated in 7 columns.
- **Ranges:** no negative or implausible values. `tenure` ranges from 0 to 72 months (72 is likely a system cap). `TotalCharges` is right-skewed (mean 2,280 vs median 1,395), so medians are used for group comparisons.

### 2. Who churns: recommendations for the Retention team
Overall churn rate: **26.5%** (1,869 of 7,043).

**Highest-risk segments:**
1. **Customers in their first year:** 47.4% churn (**1.8×** the average), and the largest group (2,186 customers). After 4+ years churn drops to 9.5%.
2. **Month-to-month contracts:** 42.7% churn (**1.6×**), vs 11.3% for one-year and 2.8% for two-year contracts.
3. **Electronic check payment:** 45.3% churn (**1.7×**), vs 15–19% for the other payment methods.

**Other notable signals:** Fiber optic internet (41.9%), no tech support (41.6%), no online security (41.8%), senior citizens (41.7%), no dependents (31.3%), paperless billing (33.6%).

**Little or no relationship with churn:** gender (26.9% vs 26.2%), phone service, multiple lines, and streaming services (~30% vs ~34% among internet users).

**Numeric features (median, churned vs retained):**
- `tenure`: **10 vs 38 months**, the strongest difference in the dataset.
- `MonthlyCharges`: **79.65 vs 64.43** (churned customers pay ~24% more per month).
- `TotalCharges`: 704 vs 1,680. This is a consequence of shorter tenure, not an independent finding.

**Caveats:**
- These segments **overlap heavily**: a first-year, month-to-month, electronic-check, fiber customer is often the same person. Segment sizes cannot be added up.
- These are **correlations, not causes**.

### 3. Features for modelling (input to MLP-1)
- **Target:** `Churn` (Yes/No → 1/0). Classes are imbalanced (26.5% positive), so use a **stratified** train/test split and evaluate with ROC AUC rather than accuracy.
- **Categorical (one-hot):** `gender`, `Partner`, `Dependents`, `PhoneService`, `MultipleLines`, `InternetService`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`, `Contract`, `PaperlessBilling`, `PaymentMethod`.
- **Binary, already encoded:** `SeniorCitizen` (0/1, pass through unchanged).
- **Numeric:** `tenure`, `MonthlyCharges`, `TotalCharges`.
- **Dropped:** `customerID` (identifier, no predictive meaning).
- **Preprocessing must live in the pipeline, not the notebook:** the API will receive raw values (e.g. `"Yes"`, `" "` in `TotalCharges`), so type conversion, filling and encoding have to be part of the saved model artifact.

**Risks to address in MLP-1:**
- **`TotalCharges` multicollinearity:** r = 0.83 with `tenure` and 0.65 with `MonthlyCharges` (≈ tenure × monthly charge). Decide whether to keep it for a logistic regression.
- **Redundant service categories:** `"No internet service"` / `"No phone service"` duplicate `InternetService` / `PhoneService`. Consider mapping them to `"No"`.
- `tenure_group` was created for EDA only and is **not** a model feature.